controller/Dockerfile

```diff
@@ -21,56 +21,31 @@ FROM --platform=$TARGETPLATFORM mcr.microsoft.com/windows/servercore@sha256:4595
 # build stages
 
 # intermediate go generate stage
-FROM golang AS intermediate 
+FROM golang AS bins
 ARG APP_INSIGHTS_ID # set to enable AI telemetry
 ARG GOARCH=amd64 # default to amd64
 ARG GOOS=linux # default to linux
 ENV CGO_ENABLED=0
 ENV GOARCH=${GOARCH}
 ENV GOOS=${GOOS}
-COPY . /go/src/github.com/microsoft/retina
-WORKDIR /go/src/github.com/microsoft/retina
 RUN if [ "$GOOS" = "linux" ] ; then \
-    tdnf install -y clang16 lld16 bpftool libbpf-devel; \
-    go generate -x /go/src/github.com/microsoft/retina/pkg/plugin/...; \
+      tdnf install -y clang16 lld16 bpftool libbpf-devel; \
     fi
-
-# capture binary
-FROM golang AS capture-bin
-ARG APP_INSIGHTS_ID # set to enable AI telemetry
-ARG GOARCH=amd64 # default to amd64
-ARG GOOS=linux # default to linux
-ARG VERSION
-ENV CGO_ENABLED=0
-ENV GOARCH=${GOARCH}
-ENV GOOS=${GOOS}
-COPY . /go/src/github.com/microsoft/retina 
+COPY ./pkg/plugin/ /go/src/github.com/microsoft/retina/plugin
 WORKDIR /go/src/github.com/microsoft/retina
-RUN --mount=type=cache,target="/root/.cache/go-build" go build -v -o /go/bin/retina/captureworkload -ldflags "-X main.version="$VERSION" -X main.applicationInsightsID="$APP_INSIGHTS_ID"" captureworkload/main.go
+RUN if [ "$GOOS" = "linux" ] ; then \
+      go mod init github.com/microsoft/retina; \
+      go generate -x /go/src/github.com/microsoft/retina/pkg/plugin/...; \
+      rm go.mod; \
+    fi
 
+COPY ./go.mod ./go.sum ./
+RUN go mod download
 
-# controller binary
-FROM intermediate AS controller-bin
-ARG APP_INSIGHTS_ID # set to enable AI telemetry
-ARG GOARCH=amd64 # default to amd64
-ARG GOOS=linux # default to linux
-ARG VERSION
-ENV CGO_ENABLED=0
-ENV GOARCH=${GOARCH}
-ENV GOOS=${GOOS}
+COPY . .
 RUN --mount=type=cache,target="/root/.cache/go-build" go build -v -o /go/bin/retina/controller -ldflags "-X main.version="$VERSION" -X main.applicationInsightsID="$APP_INSIGHTS_ID"" controller/main.go 
-
-
-# init binary
-FROM intermediate AS init-bin
-ARG APP_INSIGHTS_ID # set to enable AI telemetry
-ARG GOARCH=amd64 # default to amd64
-ARG GOOS=linux # default to linux
-ARG VERSION
-ENV CGO_ENABLED=0
-ENV GOARCH=${GOARCH}
-ENV GOOS=${GOOS}
 RUN --mount=type=cache,target="/root/.cache/go-build" go build -v -o /go/bin/retina/initretina -ldflags "-X main.version="$VERSION" -X main.applicationInsightsID="$APP_INSIGHTS_ID"" init/retina/main_linux.go
+RUN --mount=type=cache,target="/root/.cache/go-build" go build -v -o /go/bin/retina/captureworkload -ldflags "-X main.version="$VERSION" -X main.applicationInsightsID="$APP_INSIGHTS_ID"" captureworkload/main.go
```

Comment:
could this be slower since it will serially build all the binaries instead of being parallelizable separate images? if part what you're trying to improve is redundant dep downloads and rebuilds, could we move those upstream to the `golang` image and still have separate parallel final bin builds?

Reply:
@rbtr I did end up switching this. As mentioned in the commit, there wasn't a huge benefit to parallel builds (about 2s). But it wasn't worse, and hopefully the effect is compounded in CI builds.